File: 황영훈/week07/키패드_누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        int lastLeftNumber = 10; // "*"은 10으로 치환
        int lastRightNumber = 12; // "#"은 12로 치환
        String answer = "";
        for (int number : numbers) {
            // 1, 4, 7 -> 왼손 사용
            if (number == 1 || number == 4 || number == 7) {
                answer += "L";
                lastLeftNumber = number;
            }
            // 3, 6, 9 -> 오른손 사용
            else if (number == 3 ||number == 6 || number == 9) {
                answer += "R";
                lastRightNumber = number;
            }
            // 2, 5, 8, 0 -> 가까운 손 사용
            else {
                // 0은 11로 치환
                if (number == 0) {
                    number = 11;
                }
                // 가까운 손 탐색 로직 수행
                String closeHand = getCloseHand(lastLeftNumber, lastRightNumber, number, hand);
                answer += closeHand;
                if (closeHand.equals("L")) {
                    lastLeftNumber = number;
                } else {
                    lastRightNumber = number;
                }
            }
        }
        return answer;
    }
    
    private String getCloseHand(int lastLeftNumber, int lastRightNumber, int number, String hand) {
        int leftDistance = Math.abs(number - lastLeftNumber) / 3 + Math.abs(number - lastLeftNumber) % 3;
        int rightDistance = Math.abs(number - lastRightNumber) / 3 + Math.abs(number - lastRightNumber) % 3;
        if (leftDistance < rightDistance) {
            return "L";
        }
        if (leftDistance > rightDistance) {
            return "R";
        }
        return (hand.equals("left")) ? "L" : "R";
    }
}

File: 이수빈/week7/키패드누르기.py
def solution(numbers, hand):
    answer = ""
    l = 10 # 왼손의 위치
    r = 11 # 오른손의 위치
    # 각 숫자의 위치 값 저장
    number_pad = [[3, 1], 
                  [0, 0], [0, 1], [0, 2], 
                  [1, 0], [1, 1], [1, 2], 
                  [2, 0], [2, 1], [2, 2], 
                  [3, 0], [3, 2]]
    for n in numbers:
        if n in [1, 4, 7]:
            answer += "L"
            l = n
        elif n in [3, 6, 9]:
            answer += "R"
            r = n
        else: # 2, 5, 8, 0 중 하나일 경우 맨해튼 거리 (|x1−x2|+|y1−y2|)를 구하여 둘 중 짧은 쪽을 택한다
            l_dist = abs(number_pad[l][0] - number_pad[n][0]) + abs(number_pad[l][1] - number_pad[n][1])
            r_dist = abs(number_pad[r][0] - number_pad[n][0]) + abs(number_pad[r][1] - number_pad[n][1])
            if l_dist < r_dist:
                answer += "L"
                l = n
            elif l_dist > r_dist:
                answer += "R"
                r = n
            else:
                if hand == "left":
                    answer += "L"
                    l = n
                else:
                    answer += "R"
                    r = n
            
    return answer